### PR TITLE
feat: Added option to Create Event

### LINKED ISSUE #518: Add option to create Event
Add option to create an event with required fields

This issue follows creation of event with only basic bare minimum fields

### PR DESCRIPTION
**Pull request by Masquerade0097**
Fixes #518 

Changes: User can now create Event in the app with mandatory fields.

Screenshots for the change: 
brscr2

### CODE REVIEW

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/CreateEvent.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+@AllArgsConstructor
+@NoArgsConstructor
+@JsonNaming(PropertyNamingStrategy.KebabCaseStrategy.class)
+@SuppressWarnings({ "PMD.TooManyFields" })
+@JsonInclude(JsonInclude.Include.ALWAYS)
+@EqualsAndHashCode(exclude = { "creating", "deleting" })
+@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY, getterVisibility = JsonAutoDetect.Visibility.NONE,
+    setterVisibility = JsonAutoDetect.Visibility.NONE, creatorVisibility = JsonAutoDetect.Visibility.NONE,
+    isGetterVisibility = JsonAutoDetect.Visibility.NONE)
+public class CreateEvent {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/data/models/CreateEvent.java]: We don't need another model. Original model can be reused

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/CreateEventWrap.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+@JsonInclude(JsonInclude.Include.ALWAYS)
+public class CreateEventWrap {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/data/models/CreateEventWrap.java]: Same here. Please see the format of creating tickets for more info

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/CreateEvent.java (hunk 1 above)
Comment (Masquerade0097) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/CreateEvent.java]: Original model Event structure is different from the requested one, it contains many unnecessary fields and also I had to assign default values for all the fields. That's why I created another object.

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/CreateEvent.java]: This is duplicating code, we'll have to maintain two models this way. Please use the existing one. Because having extra fields poses no problem in POST request. For default fields, one static method can be created to populate those when creating a new one

Comment (Masquerade0097) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/CreateEvent.java]: @iamareebjamal   I've tried using the Event object already present but it contains id tag when JSON request is made which I'm not able to remove because it gives HTTP 400 error - `"Event id not found"` . Is there any way I can hide that id tag( `@JsonIgnore` doesn't work for it, neither does `@JsonInclude(JsonInclude.Include.NON_NULL` when setting id to null)  like currently it's being done in Ticket object. In the Ticket create request, the id tag is not present although it is present in the Object, i can't understand how it's done.

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/CreateEvent.java]: Null fields should not be read by server. Can you please post the request being generated

Comment (Masquerade0097) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/CreateEvent.java]: When id is set to null, the request doesn't generate. The event create activity crashes.

Comment (Masquerade0097) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/CreateEvent.java]: id field is not setting to null by default. This is the request JSON -
 `D/OkHttp: {"data":{"type":"event","id":"0","attributes":{"analytics":{"checkedIn":{},"donationTickets":{},"freeTickets":{},"paidTickets":{},"soldDonationTickets":{},"soldFreeTickets":{},"soldPaidTickets":{},"totalAttendees":{},"totalSale":{},"totalTickets":{}},"bank-details":"example","can-pay-by-bank":"false","can-pay-by-cheque":"false","can-pay-by-paypal":"false","can-pay-by-stripe":"false","can-pay-onsite":"false","cheque-details":"example","code-of-conduct":"example","description":"example","ends-at":"2018-01-30T16:54:56.956+05:30","event-delegate":{"header":"LIVE","headerId":2337004},"external-event-url":"http://www.example.com","has-organizer-info":"false","is-complete":false,"is-sponsors-enabled":"false","latitude":"0.0","location-name":"example","logo-url":"http://www.example.com","longitude":"0.0","name":"I fgyhuh","onsite-details":"example","organizer-description":"example","organizer-name":"example","original-image-url":"http://www.example.com","payment-country":"example","payment-currency":"INR","paypal-email":"example","privacy":"example","searchable-location-name":"example","starts-at":"2018-01-24T16:54:56.956+05:30","state":"draft","ticket-url":"http://www.example.com","timezone":"Asia/Kolkata","is-map-shown":"false","is-sessions-speakers-enabled":"false","is-tax-enabled":"false","is-ticketing-enabled":"false"}}}`

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/CreateEvent.java]: Set to null

**Comment by Masquerade0097** [comment]
@iamareebjamal Please review.

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java [hunk 3]
```diff
@@ -32,10 +44,18 @@
 @AllArgsConstructor
 @JsonNaming(PropertyNamingStrategy.KebabCaseStrategy.class)
 @Table(database = OrgaDatabase.class, allFields = true)
+@JsonInclude(JsonInclude.Include.NON_DEFAULT)
 @EqualsAndHashCode(exclude = { "eventDelegate", "analytics" })
 @SuppressWarnings({ "PMD.ExcessivePublicCount", "PMD.TooManyFields" })
+@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY, getterVisibility = JsonAutoDetect.Visibility.NONE,
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: Why is this needed?

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java [hunk 4]
```diff
@@ -32,10 +44,18 @@
 @AllArgsConstructor
 @JsonNaming(PropertyNamingStrategy.KebabCaseStrategy.class)
 @Table(database = OrgaDatabase.class, allFields = true)
+@JsonInclude(JsonInclude.Include.NON_DEFAULT)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: Why is this needed?

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java [hunk 5]
```diff
@@ -47,55 +67,84 @@
...
+    public boolean isSessionsSpeakersEnabled;
+    @JsonProperty("is-ticketing-enabled")
+    @JsonSerialize(using = ToStringSerializer.class)
+    public boolean isTicketingEnabled;
+    @JsonProperty("is-tax-enabled")
+    @JsonSerialize(using = ToStringSerializer.class)
+    public boolean isTaxEnabled;
+    @JsonProperty("is-map-shown")
+    @JsonSerialize(using = ToStringSerializer.class)
+    public boolean isMapShown;
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: This shouldn't be serialized to string. If this is a requirement, it is more likely a bug on server

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java [hunk 6]
```diff
@@ -112,4 +161,25 @@ public Event() { }
...
+        state = STATE_DRAFT;
+        searchableLocationName = DEFAULT_STRING;
+        description = DEFAULT_STRING;
+        bankDetails = DEFAULT_STRING;
+        chequeDetails = DEFAULT_STRING;
+        originalImageUrl = DEFAULT_PROTOCOL + File.separator + File.separator + DEFAULT_URL;
+        externalEventUrl = DEFAULT_PROTOCOL + File.separator + File.separator + DEFAULT_URL;
+        logoUrl = DEFAULT_PROTOCOL + File.separator + File.separator + DEFAULT_URL;
+        ticketUrl = DEFAULT_PROTOCOL + File.separator + File.separator + DEFAULT_URL;
+    }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: By single responsibility principle, this should not be in this class at all, this class is model which should be mostly untouched

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: Also, don't initialize by default. These fields should be null if not provided by user

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java [hunk 7]
```diff
@@ -32,10 +44,18 @@
...
 @SuppressWarnings({ "PMD.ExcessivePublicCount", "PMD.TooManyFields" })
+@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY, getterVisibility = JsonAutoDetect.Visibility.NONE,
+    setterVisibility = JsonAutoDetect.Visibility.NONE, creatorVisibility = JsonAutoDetect.Visibility.NONE,
+    isGetterVisibility = JsonAutoDetect.Visibility.NONE)
 public class Event implements Comparable<Event>, IHeaderProvider {
 
+    private static final String DEFAULT_STRING = "example";
+    private static final String DEFAULT_PROTOCOL = "http:";
+    private static final String DEFAULT_URL = "www.example.com";
+    private static final String DEFAULT_CURRENCY = "INR";
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: See comment below about single responsibility

File: app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventPresenter.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    @Inject
+    public CreateEventPresenter(IEventRepository eventRepository) {
+        this.eventRepository = eventRepository;
+        LocalDateTime current = LocalDateTime.now();
+
+        String isoDate = DateUtils.formatDateToIso(current);
+        event.getStartsAt().set(isoDate);
+        event.getEndsAt().set(isoDate);
+        event.setDefaults();
+        event.setId(null);
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventPresenter.java]: It's null by default

File: app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventPresenter.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    public void createEvent() {
+        if (!verify())
+            return;
+
+        eventRepository
+            .createEvent(event)
+            .compose(dispose(getDisposable()))
+            .compose(erroneous(getView()))
+            .doOnSubscribe(disposable -> event.creating.set(true))
+            .doFinally(() -> event.creating.set(false))
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventPresenter.java]: Rather than using event observable, use Rx transformers to do this and the view should be already implementing correct methods to show or hide the progress

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java (hunk 4 above)
Comment (Masquerade0097) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: This was added to reject the fields with non default values such that the request json doesn't contain unwanted stuff but it don't affect the functionality of the app i.e. event can be created without it. So i'll remove it.

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java (hunk 3 above)
Comment (Masquerade0097) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: There are some fields like `isMapShown` which by default are getting serialized to two fields - `is-map-shown` and `map-shown`. So to prevent this I added the above annotation after which it as only getting serialized to `is-map-shown`.

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: Yes

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java (hunk 5 above)
Comment (Masquerade0097) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: Yes I have to add them. Please see my above comment.

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java (hunk 6 above)
Comment (Masquerade0097) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: When the fields are not initialized by default they are having null values which in turn causes NullPointerException in Dashboard which in turn shows blank. Also just after creating event I get a whte screen and app hangs though the event is created successfully. But on choosing that event in the event list layout, again the app hangs.  This is the error log for NullPointerException - 
`E/Logger: An exception occurred : value is null
          java.lang.NullPointerException: value is null
              at io.reactivex.internal.functions.ObjectHelper.requireNonNull(ObjectHelper.java:39)
              at io.reactivex.Single.just(Single.java:655)
              at org.fossasia.openevent.app.common.utils.core.CurrencyUtils.getCurrencySymbol(CurrencyUtils.java:44)
              at org.fossasia.openevent.app.module.main.MainPresenter.lambda$start$1$MainPresenter(MainPresenter.java:68)
              at org.fossasia.openevent.app.module.main.MainPresenter$$Lambda$2.accept(Unknown Source)
              at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
              at io.reactivex.internal.operators.observable.ObservableDoOnEach$DoOnEachObserver.onNext(ObservableDoOnEach.java:103)
              at io.reactivex.internal.operators.observable.ObservableDoOnEach$DoOnEachObserver.onNext(ObservableDoOnEach.java:103)
              at io.reactivex.internal.observers.DisposableLambdaObserver.onNext(DisposableLambdaObserver.java:58)
              at io.reactivex.subjects.PublishSubject$PublishDisposable.onNext(PublishSubject.java:264)
              at io.reactivex.subjects.PublishSubject.onNext(PublishSubject.java:183)
              at org.fossasia.openevent.app.common.data.Bus.pushSelectedEvent(Bus.java:19)
              at org.fossasia.openevent.app.module.event.list.EventsListAdapter$EventRecyclerViewHolder.lambda$new$0$EventsListAdapter$EventRecyclerViewHolder(EventsListAdapter.java:76)
              at org.fossasia.openevent.app.module.event.list.EventsListAdapter$EventRecyclerViewHolder$$Lambda$0.onClick(Unknown Source)
              at android.view.View.performClick(View.java:5612)
              at android.view.View$PerformClick.run(View.java:22285)
              at android.os.Handler.handleCallback(Handler.java:751)
              at android.os.Handler.dispatchMessage(Handler.java:95)
              at android.os.Looper.loop(Looper.java:154)
              at android.app.ActivityThread.main(ActivityThread.java:6123)
              at java.lang.reflect.Method.invoke(Native Method)
              at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:867)
              at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:757)`

File: app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventPresenter.java (hunk 8 above)
Comment (Masquerade0097) [reply app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventPresenter.java]: Ya, I'll remove it.

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java (hunk 7 above)
Comment (Masquerade0097) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: Please read my comment below about setting default values. If needed, I'll initialize them in the presenter.

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java (hunk 5 above)
Comment (Masquerade0097) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: Earlier i was getting server error sucht that it was accpting `"false"` and not `false` that's why i had to add it but now I'm not getting any related errors. So I'll remove it then.

File: app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventPresenter.java (hunk 9 above)
Comment (Masquerade0097) [reply app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventPresenter.java]: Ok

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java (hunk 6 above)
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: Then ask for the payment currency from the user. And please create an issue to handle null currency type

**Comment by Masquerade0097** [comment]
@iamareebjamal I've made the changes, please have a look.

**Comment by Masquerade0097** [comment]
@iamareebjamal I've made the changes, please have a look.

File: app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java [hunk 10]
```diff
@@ -47,54 +52,58 @@
...
+    public boolean isTicketingEnabled;
+    public boolean isTaxEnabled;
+    public boolean isMapShown;
+
+    @JsonSerialize(using = ObservableStringSerializer.class)
+    @JsonDeserialize(using = ObservableStringDeserializer.class)
+    public ObservableString startsAt = new ObservableString();
+    @JsonSerialize(using = ObservableStringSerializer.class)
+    @JsonDeserialize(using = ObservableStringDeserializer.class)
+    public ObservableString endsAt = new ObservableString();
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: This is a deprecated method of binding and warnings are displayed. Data binding developers will remove this method soon, they have advised to not use it

Comment (Masquerade0097) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: @iamareebjamal CreateEvent as well as AboutEvent uses this method of binding and gives a warning. Both use the same DateService methods etc can I open a separate issue for this warning and use the same mechanism for both of them because changing them in one place will give error in other. What do you say?

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/app/common/data/models/Event.java]: Yup :+1:
